src/mumble/GlobalShortcut_win.h

```diff
@@ -6,135 +6,94 @@
-	static BOOL CALLBACK EnumDevicesCB(LPCDIDEVICEINSTANCE, LPVOID);
-	static BOOL CALLBACK EnumDeviceObjectsCallback(LPCDIDEVICEOBJECTINSTANCE lpddoi, LPVOID pvRef);
-	static LRESULT CALLBACK HookKeyboard(int, WPARAM, LPARAM);
-	static LRESULT CALLBACK HookMouse(int, WPARAM, LPARAM);
-
-	/// Handle an incoming Windows keyboard message.
+	/// Inject a native Windows raw input message into GlobalShortcutWin's
+	/// thread. This method is meant to be called from the main thread
+	/// to pass native Windows keyboard messages to GlobalShortcutWin.
 	///
-	/// Returns true if the GlobalShortcut engine signals that the
-	/// button should be suppressed. Returns false otherwise.
-	static bool handleKeyboardMessage(DWORD scancode, DWORD vkcode, bool extended, bool down);
+	/// @param msg The raw input handle to inject into GlobalShortcutWin.
+	void injectRawInputMessage(HRAWINPUT handle);
 
-	/// Handle an incoming Windows mouse message.
-	///
-	/// Returns true if the GlobalShortcut engine signals that the
-	/// button should be suppressed. Returns false otherwise.
-	static bool handleMouseMessage(unsigned int btn, bool down);
+	ButtonInfo buttonInfo(const QVariant &button) override;
+	void run() override;
 
-	virtual bool canSuppress() Q_DECL_OVERRIDE;
-	void run() Q_DECL_OVERRIDE;
-	bool event(QEvent *e) Q_DECL_OVERRIDE;
+	GlobalShortcutWin();
+	~GlobalShortcutWin() override;
 public slots:
+	void deviceRemoved(const HANDLE deviceHandle);
 	void timeTicked();
 
-public:
-	GlobalShortcutWin();
-	~GlobalShortcutWin() Q_DECL_OVERRIDE;
-	void unacquire();
-	ButtonInfo buttonInfo(const QVariant &) Q_DECL_OVERRIDE;
-
-	/// Inject a native Windows keyboard message into GlobalShortcutWin's
-	/// event stream. This method is meant to be called from the main thread
-	/// to pass native Windows keyboard messages to GlobalShortcutWin.
-	///
-	/// @param  msg  The keyboard message to inject into GlobalShortcutWin.
-	///              Must be WM_KEYDOWN, WM_KEYUP, WM_SYSKEYDOWN or WM_SYSKEYUP.
-	///              Otherwise the message will be ignored.
-	///
-	/// @return      Returns true if the GlobalShortcut engine signalled that
-	///              the button should be suppressed. Returns false otherwise.
-	bool injectKeyboardMessage(MSG *msg);
+	void on_hidMessage(const HANDLE deviceHandle, std::vector< char > reports, const uint32_t reportSize);
+	void on_keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey);
+	void on_mouseMessage(const uint16_t flags, const uint16_t data);
+signals:
+	void hidMessage(const HANDLE deviceHandle, std::vector< char > reports, const uint32_t reportSize);
+	void keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey);
+	void mouseMessage(const uint16_t flags, const uint16_t data);
+
+protected:
+	struct Device {
+		std::string name;
+		std::string prefix;
+		std::vector< uint8_t > data;
+		std::vector< bool > buttons;
```

Comment:
Note that
> The storage is not necessarily an array of bool values, but the library implementation may optimize storage so that each value is stored in a single bit.

[source](https://www.cplusplus.com/reference/vector/vector-bool/)

Reply:
Good catch, I wasn't aware of that!

As long as we're not treating the internal buffer as a contiguous array there should be no issues though. From https://en.cppreference.com/w/cpp/container/vector_bool:

> Exposes class `std::vector<bool>::reference` as a method of accessing individual bits. In particular, objects of this class are returned by [operator[]](https://en.cppreference.com/w/cpp/container/vector/operator_at) by value.

Reply:
> As long as we're not treating the internal buffer as a contiguous array there should be no issues though

Yes indeed. I just wanted to make sure you are aware of this and that this usage is therefore not required.